packages/eslint-plugin/src/rules/prefer-return-this-type.ts

```diff
@@ -0,0 +1,161 @@
+        return false;
+      }
+
+      const func = parserServices.esTreeNodeToTSNodeMap.get(originalFunc);
+
+      // two things to note here:
+      // 1. arrow function without brackets is flagged as HasImplicitReturn by ts.
+      // 2. ts.NodeFlags.HasImplicitReturn is not accurate. TypeScript compiler uses control flow
+      //    analysis to determine if a function has implicit return.
+      const hasImplicitReturn =
+        func.flags & ts.NodeFlags.HasImplicitReturn &&
+        !(
+          func.kind === ts.SyntaxKind.ArrowFunction &&
+          func.body?.kind !== ts.SyntaxKind.Block
+        );
+
+      if (hasImplicitReturn || !func.body) {
+        return false;
+      }
+
+      const classType = checker.getTypeAtLocation(
+        parserServices.esTreeNodeToTSNodeMap.get(originalClass),
+      ) as ts.InterfaceType;
+
+      if (func.body.kind !== ts.SyntaxKind.Block) {
+        const type = checker.getTypeAtLocation(func.body);
+        return classType.thisType === type;
+      }
+
+      let alwaysReturnsThis = true;
+      let hasReturnStatements = false;
+
+      forEachReturnStatement(func.body as ts.Block, stmt => {
+        hasReturnStatements = true;
+
+        const expr = stmt.expression;
+        if (!expr) {
+          alwaysReturnsThis = false;
+          return true;
+        }
+
+        // fast check
+        if (expr.kind === ts.SyntaxKind.ThisKeyword) {
+          return;
+        }
+
+        const type = checker.getTypeAtLocation(expr);
+        if (classType.thisType !== type) {
+          alwaysReturnsThis = false;
+          return true;
+        }
+
+        return undefined;
+      });
+
+      return hasReturnStatements && alwaysReturnsThis;
+    }
+
+    function checkFunction(
+      originalFunc: FunctionLike,
+      originalClass: ClassLikeDeclaration,
+    ): void {
+      if (
+        !originalFunc.returnType ||
+        IgnoreTypes.has(originalFunc.returnType.typeAnnotation.type)
+      ) {
+        return;
+      }
+
+      if (isFunctionAlwaysReturningThis(originalFunc, originalClass)) {
+        context.report({
+          node: originalFunc.returnType,
+          messageId: 'UseThisType',
+          fix(fixer) {
+            return fixer.replaceText(
+              originalFunc.returnType!.typeAnnotation,
+              'this',
+            );
+          },
+        });
+      }
+    }
+
+    return {
+      'ClassBody > MethodDefinition'(node: TSESTree.MethodDefinition): void {
+        checkFunction(node.value, node.parent!.parent as ClassLikeDeclaration);
+      },
+      'ClassBody > ClassProperty'(node: TSESTree.ClassProperty): void {
+        if (
+          !(
+            node.value?.type === AST_NODE_TYPES.FunctionExpression ||
+            node.value?.type === AST_NODE_TYPES.ArrowFunctionExpression
+          )
+        ) {
+          return;
+        }
+
+        checkFunction(node.value, node.parent!.parent as ClassLikeDeclaration);
+      },
```

Comment:
one easier way to do this is to use a stack to track your state.
It saves you doing manual traversal of the tree

```ts
interface Stack { classLike: ClassLikeDeclaration, methodLike: FunctionLike | null }
const stack: Stack[] = [];
function getCurrentStack(): Stack | undefined {
  return stack[stack.length - 1];
}

return {
  'ClassDeclaration, ClassExpression:enter'(node: ClassLikeDeclaration) {
    stack.push({ classLike: node, methodLike: null });
  },
  'ClassDeclaration, ClassExpression:exit'(node: ClassLikeDeclaration) {
    stack.pop();
  },
  'ClassBody > MethodDefinition:enter, ClassBody > ClassProperty > :matches(FunctionExpression, ArrowFunctionExpression).value:enter'(node: FunctionLike ) {
    const current = getCurrentStack();
    if (current != null) {
      current.methodLike = node;
    }
  },
  'ClassBody > MethodDefinition, ClassBody > ClassProperty > :matches(FunctionExpression, ArrowFunctionExpression).value:exit'(node: FunctionLike ) {
    const current = getCurrentStack();
    if (current != null) {
      current.methodLike = null;
    }
  },
  ReturnStatement(node) {
    const current = getCurrentStack();
    if (current != null && current.methodLike != null) {
      // check return and report
    }
  },
};
```

Reply:
Cool, I haven't seen this pattern before, will try it.

Reply:
This is more complicated than I have expected. We may have functions inside methods and need to track them too.

```ts
class Foo {
  bar() {
    const f = function() {
       return this;
    }
  }
}
```